### DISCUSSION TITLE
Couldn't find any assemblies on NuGet package `CefSharp.Common.NETCore` in both 99.2.140 and 100.0.140

### DISCUSSION THREAD

**Discussion by kekyo** [cefsharp/CefSharp#4061] [General]
I tried to write a small verification code using the .NET Core version of CefSharp, but could not find any symbols in the CefSharp initialization code. It is the following startup code, which builds fine with `net48` instead `netcoreapp3.1`.

```csharp
var settings = new CefSettings();
Cef.Initialize(settings, performDependencyCheck: true, browserProcessHandler: null);
```

I checked inside the package and I can't find the `lib` directory, so I assume the problem is related to packaging.

* [NuGet package explorer in .NET Core 100.0.140](https://nuget.info/packages/CefSharp.Common.NETCore/100.0.140)
* [NuGet package explorer in .NET Framework 100.0.140](https://nuget.info/packages/CefSharp.Common/100.0.140)

Thanks for a very good project.

**Comment by amaitland** [comment]
Firstly please use discussions to ask questions. 

> but could not find any symbols in the CefSharp initialization code. 

What do you mean by this? 

What exactly are you trying to achieve?

**Comment by kekyo** [comment]
> What do you mean by this?

I wanted to do a quick run through of AnyCPU support in CefSharp (.NET Core version), so I referred to https://github.com/cefsharp/CefSharp/issues/1714#issuecomment-227041722 I tried to insert the code:

* If the TargetFramework is `net48` and the package is `CefSharp.Wpf`, it builds and works OK.
* If the TargetFramework is `netcoreapp3.1` and the package is `CefSharp.Wpf.NETCore`, I get a C# compiler error that the `CefSettings` and `Cef` symbols are missing.

So I looked for where these symbols are defined in the `CefSharp.Wpf` package and found them in the `CefSharp.Common` package. But in `CefSharp.Common.NETCore` package is not included.

You can see it in the attached NuGet package explorer link.

**Reply by amaitland** [reply]
There are details about how it's packaged in #2796 a runtimes folder is used. .Net core is packaged differently. AnyCPU is only meant for libraries for .net core. 

If you wish to support. Net Core and .Net 4.8 the  you need to conditionally add the packages. There's an example you can use at https://github.com/cefsharp/CefSharp.Wpf.HwndHost/blob/master/CefSharp.Wpf.HwndHost/CefSharp.Wpf.HwndHost.csproj#L25

There is a generic working example at https://github.com/cefsharp/CefSharp.MinimalExample

**Comment by kekyo** [comment]
Not knowing if the problem is not shared or if my understanding is incorrect, I created a very small WPF application.

The project is multi-built using `TargetFrameworks` so that it can build both `net48` and `netcoreapp3.1`. Naturally, since the packages to be referenced are different, I have written them to change with `Condition`.

https://github.com/kekyo/CefSharpNETCoreTest/blob/main/CefSharpNETCoreTest/CefSharpNETCoreTest.csproj

If `TargetFrameworks` is set to only `net48` in this way, the build will succeed, but if `netcoreapp3.1` or both are set, the above mentioned compilation error will occur.

Couldn't it maybe be possible to use CefSharp to build this way?

**Reply by kekyo** [reply]
Enabled `netcoreapp3.1`:

```
Rebuild started...
Restored D:\Projects\CefSharpNETCoreTest\CefSharpNETCoreTest\CefSharpNETCoreTest.csproj (in 39 ms).
1>------ Rebuild All started: Project: CefSharpNETCoreTest, Configuration: Debug Any CPU ------
1>D:\Projects\CefSharpNETCoreTest\CefSharpNETCoreTest\App.xaml.cs(17,13,17,23): error CS0103: The name 'CefRuntime' does not exist in the current context
1>Done building project "CefSharpNETCoreTest.csproj" -- FAILED.
1>CefSharpNETCoreTest -> D:\Projects\CefSharpNETCoreTest\CefSharpNETCoreTest\bin\Debug\net48\CefSharpNETCoreTest.exe
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```

Mentioned location is here: https://github.com/kekyo/CefSharpNETCoreTest/blob/main/CefSharpNETCoreTest/App.xaml.cs#L17

**Reply by amaitland** [reply]
In future please include any error you are seeing in your initial question. 

The CefRuntime class is only in the older. Net 4.x packages. .Net Core handles dependency resolution totally differently. 

https://github.com/cefsharp/CefSharp/blob/master/CefSharp/CefSharp.netcore.csproj#L49

You need to conditionally compile the call to CefRuntime. https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives#conditional-compilation

**Reply by kekyo** [reply]
Sorry, the first error `CefSettings` and `Cef` not found, is more complex than the minimal code I rewrote above. That may have caused the error on this line as well.

I checked again and found that the only error was `CefRuntime` not found, just like the error that occurred in the minimal code above. So I had written code to use `CefRuntime`, which I had added by referring to [previous issues #1714](https://github.com/cefsharp/CefSharp/issues/1714#issuecomment-227041722) with the intention of doing `AnyCPU` support. However:

> The CefRuntime class is only in the older. Net 4.x packages. .Net Core handles dependency resolution totally differently.
> You need to conditionally compile the call to CefRuntime.

So does this mean that for `netcoreapp3.1`, the `CefRuntime` code is not needed? As shown below:

```csharp
        // https://github.com/cefsharp/CefSharp/issues/1714

        public App()
        {
#if NETFRAMEWORK
            // Have to invokes only .NET Framework environment.
            CefRuntime.SubscribeAnyCpuAssemblyResolver();
#endif

            //Any CefSharp references have to be in another method with NonInlining
            // attribute so the assembly rolver has time to do it's thing.
            InitializeCefSharp();
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        private static void InitializeCefSharp()
        {
            var settings = new CefSettings();

            Cef.Initialize(settings, performDependencyCheck: true, browserProcessHandler: null);
        }
```

Is my understanding correct? When I tested it, it did indeed compile and work successfully, but is this code okay as a way to support AnyCPU?

**Reply by amaitland** [reply]
> Is my understanding correct? When I tested it, it did indeed compile and work successfully, but is this code okay as a way to support AnyCPU?

Looks fine to me.

**Reply by kekyo** [reply]
Thank you your advice!
